File: sections/97_digit_broken_profile_dp/problems/a_digit_sum_range/solution.py
import sys
MAX_SUM = 171
WAYS = [[0] * (MAX_SUM + 1) for _ in range(20)]
WAYS[0][0] = 1
for length in range(1, 20):
    for total in range(MAX_SUM + 1):
        WAYS[length][total] = sum((WAYS[length - 1][total - digit] for digit in range(10) if digit <= total))

def count(n, target):
    if n < 0:
        return 0
    digits = list(map(int, str(n)))
    answer = 0
    prefix = 0
    for pos, limit in enumerate(digits):
        remaining = len(digits) - pos - 1
        for digit in range(limit):
            needed = target - prefix - digit
            if 0 <= needed <= MAX_SUM:
                answer += WAYS[remaining][needed]
        prefix += limit
    return answer + int(prefix == target)

def main():
    d = list(map(int, sys.stdin.buffer.read().split()))
    print('\n'.join((str(count(d[i + 1], d[i + 2]) - count(d[i] - 1, d[i + 2])) for i in range(1, len(d), 3))))
if __name__ == '__main__':
    main()

File: sections/97_digit_broken_profile_dp/problems/a_digit_sum_range/solution.cpp
#include <bits/stdc++.h>
using namespace std;
using ll=long long;
ll ways[20][172];
ll count_to(ll n,int target) {
    if(n<0)return 0;
    string digits=to_string(n);
    ll answer=0;
    int prefix=0;
    for(int pos=0;pos<(int)digits.size();pos++) {
        int limit=digits[pos]-'0',remaining=digits.size()-pos-1;
        for(int digit=0;digit<limit;digit++) {
            int needed=target-prefix-digit;
            if(0<=needed&&needed<=171)answer+=ways[remaining][needed];
        }
        prefix+=limit;
    }
    return answer+(prefix==target);
}
int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    ways[0][0]=1;
    for(int length=1;length<20;length++)for(int total=0;total<=171;total++)for(int digit=0;digit<=9&&digit<=total;digit++)ways[length][total]+=ways[length-1][total-digit];
    int q;
    cin>>q;
    while(q--) {
        ll l,r;
        int target;
        cin>>l>>r>>target;
        cout<<count_to(r,target)-count_to(l-1,target)<<'\n';
    }
}
